src/Hocon.Configuration/Config.cs

```diff
@@ -144,14 +137,10 @@ public virtual Config WithFallback(Config fallback)
         {
             if (fallback == this)
                 throw new ArgumentException("Config can not have itself as fallback", nameof(fallback));
-
-            var clone = Copy();
-
-            var current = clone;
-            while (current.Fallback != null) current = current.Fallback;
-            current.Fallback = fallback.Copy();
-
-            return clone;
+            
+            // If Fallback is not set - we will set it in new copy
+            // If Fallback was set - just use it, but with adding new fallback values
+            return new Config((HoconValue) Value.Clone(null), Fallback?.WithFallback(fallback) ?? fallback);
```

Comment:
Turned out to be much more elegant solution for getting new config with fallback, and without internal state mutations.